src/Connection/DefaultConnectionPool.php

```diff
@@ -100,23 +100,16 @@ public function getStream(Request $request, CancellationToken $cancellation): Pr
 
                 \assert($connection instanceof Connection);
 
-                if ($connection->isBusy()) {
+                if ($connection->hasStreamAvailable()) {
                     continue; // Connection is currently used to full capacity.
                 }
 
                 if (!\array_intersect($request->getProtocolVersions(), $connection->getProtocolVersions())) {
                     continue; // Connection does not support any of the requested protocol versions.
                 }
 
-                if ($connection instanceof Http2Connection && $connection->isIdle() && !(yield $connection->ping())) {
-                    continue; // Connection closed while idle.
-                }
-
-                if ($connection instanceof Http1Connection
-                    && $connection->getRemainingTime() < $this->timeoutGracePeriod
-                    && !$request->isIdempotent()
-                ) {
-                    continue; // Connection is at high-risk of closing before the request can be sent.
+                if (!yield $connection->checkLiveliness()) {
```

Comment:
Should this check be limited to non-idempotent requests?

Reply:
Maybe just pass the request in and let the connection decide?

Reply:
Yeah, that would work. `checkLiveliness` is maybe the wrong name? Have a succinct suggestion instead of `isConnectionSuitedForRequest`?

Reply:
isReuseAllowed?